src/bindgen/library.rs

```diff
@@ -443,4 +464,115 @@ impl Library {
             x.mangle_paths(&monomorphs);
         }
     }
+
+    // Replace all associated types with concrete types
+    fn replace_assoc_types(&mut self) {
+        let assoc_map = &self.assoc_types;
+
+        fn check_generic_params(generic_params: &mut GenericParams, assoc_map: &HashMap<AssocTypeId, Type>){
+            for generic_param in generic_params.0.iter_mut(){
+                if let GenericParamType::Const(ty_) = generic_param.ty() {
+                    replace_concrete(ty_, assoc_map);
+                }
+            }
+        }
+
+        // Replace associated types in constants
+        self.constants.for_all_items_mut(|const_| {
+            replace_concrete(&mut const_.ty, assoc_map);
+        });
+
+        // Replace associated types in 'static's
+        self.globals.for_all_items_mut(|static_| {
+            replace_concrete(&mut static_.ty, assoc_map);
+        });
+
+        // Replace associated types in enums
+        self.enums.for_all_items_mut(|enum_| {
+            for variant in enum_.variants.iter_mut() {
+                if let VariantBody::Body { body, .. } = &mut variant.body {
+                    for field in body.fields.iter_mut() {
+                        replace_concrete(&mut field.ty, assoc_map);
+                    }
+                    for assoc_const in body.associated_constants.iter_mut(){
+                        replace_concrete(&mut assoc_const.ty, assoc_map);
+                    }
+                    check_generic_params(&mut body.generic_params, assoc_map);
+                }
+            }
+
+            check_generic_params(&mut enum_.generic_params, assoc_map);
+        });
+
+        // Replace associated types in structs
+        self.structs.for_all_items_mut(|struct_| {
+            for field in struct_.fields.iter_mut() {
+                replace_concrete(&mut field.ty, assoc_map);
+            }
+            for assoc_const in struct_.associated_constants.iter_mut(){
+                replace_concrete(&mut assoc_const.ty, assoc_map);
+            }
+            check_generic_params(&mut struct_.generic_params, assoc_map)
+        });
```

Comment:
can we do something like this:

```suggestion
        self.structs.for_all_items_mut(|struct_| {
            struct_.resolve_assoc_types(assoc_map);
        )}
```

Reply:
likewise for others

Reply:
I would make it a method on the `Item` trait

Reply:
I didn't use Item trait in the end. Item covers all  the ir types except Function. It's strange but I can't figure out why

Reply:
so why not use `Item` for all others and implement inherent method for function?

Reply:
that is exactly how it's done with `resolve_declaration_types`